cassiopeia/solver/SpectralNeighborJoiningSolver.py

```diff
@@ -0,0 +1,318 @@
+            metric given by weighted_hamming_distance from
+            dissimilarity_functions without weights given.
+
+        add_root: Whether or not to add an implicit root to the tree, i.e. a
+            root with unmutated characters.
+        prior_transformation: Function to use when transforming
+            priors into weights
+    """
+
+    def __init__(
+        self,
+        similarity_function: Optional[
+            Callable[
+                [np.ndarray, np.ndarray, int, Dict[int, Dict[int, float]]],
+                float,
+            ]
+        ] = dissimilarity_functions.exponential_negative_hamming_distance,
+        # type: ignore
+        add_root: bool = False,
+        prior_transformation: str = "negative_log",
+    ):
+        super().__init__(
+            dissimilarity_function=similarity_function,
+            add_root=add_root,
+            prior_transformation=prior_transformation,
+        )  # type: ignore
+
+    def get_dissimilarity_map(
+        self, cassiopeia_tree: CassiopeiaTree, layer: Optional[str] = None
+    ) -> pd.DataFrame:
+        """Outputs the first lambda matrix, where every subset is an
+        individual node.
+
+        Args:
+            cassiopeia_tree: the CassiopeiaTree object passed into solve()
+            layer: Layer storing the character matrix for solving. If None, the
+                default character matrix is used in the CassiopeiaTree.
+
+        Returns:
+            DataFrame object of the lambda matrix, but similar in structure to
+                DistanceSolver's dissimilarity_map.
+        """
+
+        # get dissimilarity map and save it as private instance variable
+        self._similarity_map = super().get_dissimilarity_map(
+            cassiopeia_tree, layer
+        )
+
+        # prepare for first (pairwise) lambda matrix
+        N = self._similarity_map.shape[0]
+        node_names: np.ndarray = self._similarity_map.index.values
+        self.lambda_indices = [[i] for i in range(N)]
+
+        # generate the lambda matrix
+        lambda_matrix_arr = np.zeros([N, N])
+        for (j_idx, i_idx) in itertools.combinations(range(N), 2):
+
+            svd2_val = self._compute_svd2(
+                pair=(i_idx, j_idx), lambda_indices=self.lambda_indices
+            )
+
+            lambda_matrix_arr[i_idx, j_idx] = lambda_matrix_arr[
+                j_idx, i_idx
+            ] = svd2_val
+
+        np.fill_diagonal(lambda_matrix_arr, np.inf)
+
+        # convert array to dataframe
+        lambda_matrix_df = pd.DataFrame(
+            lambda_matrix_arr, index=node_names, columns=node_names
+        )
+
+        return lambda_matrix_df
+
+    def find_cherry(self, dissimilarity_map: np.ndarray) -> Tuple[int, int]:
+        """Finds a pair of samples to join into a cherry.
+
+        With dissimilarity_map being the lambda matrix, this method finds the
+            argmin pair of subsets of the lambda matrix.
```

Comment:
You don't have to indent the second line here -- only in the Args, Raises, and Returns sections below.